server/broadcast.go

```diff
@@ -0,0 +1,214 @@
+
+	if n.OrchestratorPool == nil {
+		glog.Info("No orchestrators specified; not transcoding")
+		return nil, ErrDiscovery
+	}
+
+	rpcBcast := core.NewBroadcaster(n)
+
+	tinfos, err := n.OrchestratorPool.GetOrchestrators(1)
+	if len(tinfos) <= 0 {
+		return nil, ErrNoOrchs
+	}
+	if err != nil {
+		return nil, err
+	}
+	tinfo := tinfos[0]
+
+	// set OSes
+	var orchOS drivers.OSSession
+	if len(tinfo.Storage) > 0 {
+		orchOS = drivers.NewSession(tinfo.Storage[0])
+	}
+
+	return &BroadcastSession{
+		Broadcaster:      rpcBcast,
+		ManifestID:       cpl.ManifestID(),
+		Profiles:         BroadcastJobVideoProfiles,
+		OrchestratorInfo: tinfo,
+		OrchestratorOS:   orchOS,
+		BroadcasterOS:    cpl.GetOSSession(),
+	}, nil
+}
+
+func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
+
+	nonce := cxn.nonce
+	rtmpStrm := cxn.stream
+	sess := cxn.sess
+	cpl := cxn.pl
+	mid := rtmpManifestID(rtmpStrm)
+	vProfile := cxn.profile
+
+	if monitor.Enabled {
+		monitor.LogSegmentEmerged(nonce, seg.SeqNo)
+	}
+
+	seg.Name = "" // hijack seg.Name to convey the uploaded URI
+	name := fmt.Sprintf("%s/%d.ts", vProfile.Name, seg.SeqNo)
+	uri, err := cpl.GetOSSession().SaveData(name, seg.Data)
+	if err != nil {
+		glog.Errorf("Error saving segment %d: %v", seg.SeqNo, err)
+		if monitor.Enabled {
+			monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+		}
+		return
+	}
+	if cpl.GetOSSession().IsExternal() {
+		seg.Name = uri // hijack seg.Name to convey the uploaded URI
+	}
+	err = cpl.InsertHLSSegment(vProfile, seg.SeqNo, uri, seg.Duration)
+	if monitor.Enabled {
+		monitor.LogSourceSegmentAppeared(nonce, seg.SeqNo, string(mid), vProfile.Name)
+		glog.V(6).Infof("Appeared segment %d", seg.SeqNo)
+	}
+	if err != nil {
+		glog.Errorf("Error inserting segment %d: %v", seg.SeqNo, err)
+		if monitor.Enabled {
+			monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+		}
+	}
+
+	if sess == nil {
+		return
+	}
+	go func() {
+		// storage the orchestrator prefers
+		if ios := sess.OrchestratorOS; ios != nil {
+			// XXX handle case when orch expects direct upload
+			uri, err := ios.SaveData(name, seg.Data)
+			if err != nil {
+				glog.Error("Error saving segment to OS ", err)
+				if monitor.Enabled {
+					monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+				}
+				return
+			}
+			seg.Name = uri // hijack seg.Name to convey the uploaded URI
+		}
+
+		// send segment to the orchestrator
+		glog.V(common.DEBUG).Infof("Submitting segment %d", seg.SeqNo)
+
+		res, err := SubmitSegment(sess, seg, nonce)
+		if err != nil {
+			if shouldStopStream(err) {
+				glog.Warningf("Stopping current stream due to: %v", err)
+				rtmpStrm.Close()
+			}
+			if shouldStopSession(err) {
+				cxn.needOrch <- struct{}{}
+			}
+			return
+		}
+
+		// download transcoded segments from the transcoder
+		gotErr := false // only send one error msg per segment list
```

Comment:
If different errors occur while downloading each segment in a list, wouldn't we want to know about all of them?

Reply:
IIRC it was done this way to minimize noise in metrics-reporting. Most errors are likely to be connectivity related, so if one fails, others are likely to too.